### PR TITLE
test: Add initial e2e tests

### PR DESCRIPTION
**Pull request by jimmidyson**
Blocked by #361.

Added Docker e2e tests running using the CAPI e2e test utilities, clusters being created, waiting for all nodes to become ready, etc. tests include both Calico and Cilium deployed via both ClusterResourceSet and HelmAddon strategies. we can add more tests but this is a good start.

~There is a data race in the test libraries which I don't know how to work around but it does not affect the usefulness of these tests and we can look at that later, unless someone knows what the problem is?~ Disabled race in ginkgo call, I think this is OK, it doesn't show races in deployed code only in tests.

### CODE REVIEW

File: pkg/handlers/generic/lifecycle/cni/calico/strategy_crs.go [hunk 1]
```diff
@@ -208,12 +218,10 @@ func generateTigeraOperatorConfigMap(
 	return namespacedTigeraConfigMap
 }
 
-func generateProviderCNICRS(
+func generateProviderCNIManifestsConfigMap(
```
Comment (dkoshkin) [inline pkg/handlers/generic/lifecycle/cni/calico/strategy_crs.go:221]: Were the changes here something you found with the e2e tests?

**Review by dkoshkin — APPROVED** [review]
Just some suggestions, but should merge this either way since these are super valuable! Thank you

File: pkg/handlers/generic/lifecycle/cni/calico/strategy_crs.go (hunk 1 above)
Comment (jimmidyson) [reply pkg/handlers/generic/lifecycle/cni/calico/strategy_crs.go:221]: Kind of, but not precisely. When I first started using the owner reference e2e checks, I hit an error because we make the `Cluster` the owner of the `ClusterResourceSet` for GC. We also set up the `Cluster` as the owner of the referenced `ConfigMaps`. I fixed this up by having our own custom ownership checks for the `ClusterResourceSets` but had to remove the ownership from `ConfigMaps` as these should be owned by the `ClusterResourceSets` (applied by CRS controller). s part of that I noticed we could use one of our utility funcs to create the CRS rather than it being applied in each controller separately...

So in summary... these changes were discovered via the e2e implementation, but actually didn't show up ny huge problems with what we were doing, it's just a little tidier now.

File: test/e2e/config/cre.yaml [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+variables:
+  # Default variables for the e2e test; those values could be overridden via env variables, thus
+  # allowing the same e2e config file to be re-used in different Prow jobs e.g. each one with a K8s version permutation.
+  # The following Kubernetes versions should be the latest versions with already published kindest/node images.
+  # This avoids building node images in the default case which improves the test duration significantly.
+  KUBERNETES_VERSION_MANAGEMENT: "v1.29.0"
+  KUBERNETES_VERSION: "v1.29.0"
+  KUBERNETES_VERSION_UPGRADE_FROM: "v1.28.0"
+  KUBERNETES_VERSION_UPGRADE_TO: "v1.29.0"
+  KUBERNETES_VERSION_API_UPGRADE_FROM: "v1.27.3"
```
[the remark this answers is not in the document]
Comment (jimmidyson) [reply test/e2e/config/cre.yaml:108]: Yes, but I would need to rework the bootstrap cluster creation as the upstream libraries do not allow specifying the bootstrap image repository, only the version, so it can only use valid tags from `kindest/node` image. I will try to push something upstream while also adding bootstrapping logic in here for the time being. But that will be a follow up PR, not in this one.

File: test/e2e/config/cre.yaml [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+images:
+  - name: ko.local/capi-runtime-extensions:${E2E_IMAGE_TAG}
+    loadBehavior: mustLoad
+
+providers:
+- name: cluster-api
+  type: CoreProvider
+  versions:
+  - name: v1.6.1
```
[the remark this answers is not in the document]
Comment (jimmidyson) [reply test/e2e/config/cre.yaml:14]: We could, but in future I would prefer to use the latest patch versions which will be resolvable in the next CAPI minor release by using `{go://sigs.k8s.io/cluster-api@v0.6}` (see https://github.com/kubernetes-sigs/cluster-api/blob/main/test/e2e/config/docker.yaml#L65 for an example). This will enable our e2es to always run against latest patch releases from CAPI which is important for compatibility tests imo.

So are you ok with keeping these values like this for now and updating once the CAPI 1.7 release is out?

File: test/e2e/config/cre.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (jimmidyson) [reply test/e2e/config/cre.yaml]: Opened an issue to follow up.
